### DISCUSSION TITLE
plugin :auto_validations

### DISCUSSION THREAD

**Discussion by PetrKaleta** [jeremyevans/sequel#1858] [General]
Hello,
I am following on discussion I've started here https://groups.google.com/g/sequel-talk/c/DH4WXCAWdIg/m/Y2Du3gPwAwAJ so how to disable `unique` auto validations on all models.

So far I was using this code

```ruby
Sequel::Model.plugin :subclasses do |sub_class|
  sub_class.skip_auto_validations(:unique)
end
```

But currently I've found, that on some models it doesn't work. And I have no idea why... Is there something wrong on this approach? Also there's one strange behavior on `subclasses` plugin, when using block, some sub_classes are anonymous... Where are they coming from?

**Comment by jeremyevans** [comment]
You can probably override the related method: `def (Sequel::Model).auto_validate_unique_columns; []; end`.

As to why your approach doesn't work, I'm not sure.  Maybe you are running it before creating the related subclasses?  Hard to guess why without a self contained example.

As to why some subclasses are anonymous, I'm guessing you are doing something like:  `class Foo < Sequel::Model(:table_name)`, and `Sequel::Model(:table_name)` creates an anonymous subclass, which you then create a subclass of.

**Reply by PetrKaleta** [reply]
Ah, okey, I understand why there are those anonymous classes now. I did some debugging and found this strange behavior in this code:

```ruby
Sequel::Model.plugin :subclasses do |sub_class|
  sub_class.skip_auto_validations(:unique)
end
```

So let's say I have model `Foo`. So in block above `sub_class == Foo` and `sub_class.auto_validate_unique_columns` is **nonempty**. After `sub_class.skip_auto_validations(:unique)` is applied then `sub_class.auto_validate_unique_columns` is `[]`, which is expected behavior. **But in code** after this block, it is again **nonempty**. Which is super strange.

I've checked auto_validations plugin source code and I think it will be related to this line `Plugins.after_set_dataset(self, :setup_auto_validations)`. It seems that for some reason, dataset for my `Foo` model was reseted... Maybe it has something to do with those anonymous classes, as I am using this approach `class Foo < Sequel::Model(:table_name)` or maybe some plugins?

This is how my `models.rb` looks like (there are no additional plugins on `Foo` model):

```ruby
require_relative 'db'

Sequel::Model.db = DB[:default]
Sequel::Model.plugin :string_stripper
Sequel::Model.plugin :boolean_readers
Sequel::Model.plugin :timestamps, update_on_create: true
Sequel::Model.plugin :auto_validations
Sequel::Model.plugin :tactical_eager_loading
Sequel::Model.plugin :subclasses do |sub_class|
  sub_class.skip_auto_validations(:unique)
end
Sequel::Model.plugin :empty_failure_backtraces if ENV['RACK_ENV'] == 'production'

Dir[File.join(__dir__, 'models/**/*.rb')].sort.each { |f| require f }

Sequel::Model.freeze_descendents unless ENV['RACK_ENV'] == 'development' || ENV['RACK_ENV'] == 'test'
```

**Reply by jeremyevans** [reply]
Try moving the `plugin :subclasses` call after the `Dir[File.join(__dir__, 'models/**/*.rb')].sort.each { |f| require f }` line.  You want to make sure all subclasses are already loaded before skipping the auto validations for them.
